python/ray/serve/_private/http_proxy.py

```diff
@@ -72,6 +74,62 @@
     )
 
 
+async def _handle_streaming_response(
+    asgi_response_generator: ray._raylet.StreamingObjectRefGenerator,
+    scope: Scope,
+    receive: Receive,
+    send: Send,
+) -> str:
+    """Consumes the `asgi_response_generator` and sends its data over `send`.
+
+    This function is essentially a proxy for a downstream ASGI response. The
+    passed generator is expected to return a stream of pickled ASGI messages
+    (dictionaries) that will be passed to the provided ASGI send interface.
+
+    Exception handling depends on if the first message has already been sent:
+        - if an exception happens *before* the first message, a 500 status will be sent.
+        - if an exception happens *after* the first message, the response stream will be
+          terminated.
+
+     This is because once the first message has been sent, the client has already
+     received the status code.
+
+    Returns:
+        status_code
+    """
+
+    status_code = ""
+    try:
+        async for obj_ref in asgi_response_generator:
+            asgi_messages: List[Dict[str, Any]] = pickle.loads(await obj_ref)
+            for asgi_message in asgi_messages:
+                # There must be exactly one "http.response.start" message that
+                # always contains the "status" field.
+                if not status_code:
+                    assert asgi_message["type"] == "http.response.start", (
+                        "First response message must be 'http.response.start'",
+                    )
+                    assert "status" in asgi_message, (
+                        "'http.response.start' message must contain 'status'",
+                    )
+                    status_code = str(asgi_message["status"])
+
+                await send(asgi_message)
+    except Exception as e:
```

Comment:
Nitpick: would be nice to catch the specific exception explicitly. But not a blocker for now, can be in a followup in the future.

Reply:
This is intended to be a very defensive catch-all handler (should only be hit if there's a system-level error)